### ISSUE TITLE
CPU Graph Not Showing

### ISSUE THREAD

**Issue by AhmadMansourAwaed** [trimble-oss/dba-dash#1149]
Hello David ,

Thanks for the wonderful tool , we are using it for almost 1 year .
I wanted to report something that :

CPU graph suddenly not showing for first node only however it's working fine with 2nd node 

not working in SQL 1 

![image](https://github.com/user-attachments/assets/8a8953f8-bf5f-425b-a3b5-ebdb6d33157a)

![image](https://github.com/user-attachments/assets/747b2a00-13e4-40bb-b0fe-cfcd06c2017c)

working in SQL 2 

![image](https://github.com/user-attachments/assets/fab23ef6-dfe6-4057-83df-791d723a91f6)

we are using latest : DBADash_3.15.0

Both nodes have the same configuration :  25 Core , 50 Logical CPU 

I am curious to investigate the issue for this strange behavior

**Comment by DavidWiseman** [comment]
Hi,  I think this might be related to #1144.  Can you run [this query](https://raw.githubusercontent.com/trimble-oss/dba-dash/refs/heads/main/DBADash/SQL/SQLCPU.sql) and let me know what it returns?  Do you see values over 100?

Is the server physical or virtual?  If it's virtual, what type of hypervisor or if it's running in the cloud, what cloud provider and instance type?  How many physical CPUs?

Thanks

**Comment by AhmadMansourAwaed** [comment]
Hello David ,
Yes , it's VM on Oracle cloud 

![image](https://github.com/user-attachments/assets/1b421135-a46a-4608-bab7-57952365e699)

![image](https://github.com/user-attachments/assets/c8d00c5d-b318-46aa-b737-f3e364315781)

i got this while running the SQLCPU query 

![image](https://github.com/user-attachments/assets/42c3ea7a-367f-49f7-81e2-b73c57ff4c1f)


Thanks for your quick response as well :)

**Comment by DavidWiseman** [comment]
The application passes in the `@TOP` parameter - the script needs to be edited slightly.  Can you run this:

```
DECLARE @TOP INT = 120
DECLARE @ts_now bigint 
SELECT @ts_now= cpu_ticks/(cpu_ticks/ms_ticks) 
FROM sys.dm_os_sys_info; 

SELECT TOP(@TOP) DATEADD(ms, -1 * (@ts_now - [timestamp]), GETUTCDATE()) AS [EventTime],
			SQLProcessUtilization AS [SQLProcessCPU], 
				 SystemIdle AS [SystemIdleProcess] 
FROM (SELECT record.value('(./Record/@id)[1]', 'int') AS record_id, 
		record.value('(./Record//SystemHealth/SystemIdle)[1]', 'int') 
		AS [SystemIdle], 
		record.value('(./Record//SystemHealth/ProcessUtilization)[1]', 'int') 
		AS [SQLProcessUtilization], [timestamp] 
		FROM (SELECT [timestamp], CONVERT(xml, record) AS [record] 
		FROM sys.dm_os_ring_buffers WITH (NOLOCK)
		WHERE ring_buffer_type = N'RING_BUFFER_SCHEDULER_MONITOR' 
		AND record LIKE N'%<SystemHealth>%') AS x) AS y 
ORDER BY record_id DESC OPTION (RECOMPILE);
```

If you can take a screenshot of this information from the hardware tab in DBA Dash it might be useful
![image](https://github.com/user-attachments/assets/815c3749-34e9-4d1f-92b2-0759e0ac9df4)

It looks like it might be a 2 socket system.  It's possible that the CPU utilization for the server is reporting up to 200% (double counting for each socket) instead of 100% and DBA Dash currently excludes values over 100%.  I'm not sure how to handle this as I've never seen the problem myself.  It's not as simple as dividing by the number of sockets as I have some two socket systems that don't exhibit this behaviour.  

The more information I have the better.  The results from the query will confirm if the utilization is reporting over 100% or if this is a different problem.

Thanks

**Comment by AhmadMansourAwaed** [comment]
Bear in mind that it's work for one node and not working for the other one however both of them are same configuration :)

node 1 

![image](https://github.com/user-attachments/assets/59d91815-da23-4981-a7c0-8736c2cc174c)

node 2 

![image](https://github.com/user-attachments/assets/d08d4742-6e04-4ebc-a80d-968de0ec2aa9)

**Comment by DavidWiseman** [comment]
It looks like DPA [fixed](https://support.solarwinds.com/SuccessCenter/s/article/CPU-utilization-is-not-displayed-properly-in-DPA) the issue dividing the SQL process utilization by the number of CPUs where the idle and CPU process are adding up to more than 100%.

I can apply a similar fix.  I think this will work:

```
DECLARE @TOP INT = 120

DECLARE @ts_now BIGINT 
DECLARE @PhysicalCPUs INT

SELECT	@ts_now= cpu_ticks/(cpu_ticks/ms_ticks),
		@PhysicalCPUs = cpu_count / hyperthread_ratio
FROM sys.dm_os_sys_info; 

SELECT TOP(@TOP) DATEADD(ms, -1 * (@ts_now - RB.timestamp), GETUTCDATE()) AS EventTime,
				CASE WHEN RB.SQLProcessUtilization+RB.SystemIdle>100 THEN RB.SQLProcessUtilization / @PhysicalCPUs ELSE RB.SQLProcessUtilization END AS SQLProcessCPU, 
				RB.SystemIdle AS [SystemIdleProcess] 
FROM (SELECT	record.value('(./Record/@id)[1]', 'int') AS record_id, 
				record.value('(./Record//SystemHealth/SystemIdle)[1]', 'int') AS SystemIdle, 
				record.value('(./Record//SystemHealth/ProcessUtilization)[1]', 'int') AS SQLProcessUtilization, 
				timestamp 
		FROM (SELECT [timestamp], CONVERT(xml, record) AS [record] 
		FROM sys.dm_os_ring_buffers WITH (NOLOCK)
		WHERE ring_buffer_type = N'RING_BUFFER_SCHEDULER_MONITOR' 
		AND record LIKE N'%<SystemHealth>%') AS x) AS RB 
ORDER BY RB.record_id DESC 
OPTION (RECOMPILE);
```

If you can try it out and let me know if it appears to be working.  The system idle and SQL process shouldn't exceed 100 with any remainder associated with non-SQL Server processes.

**Comment by AhmadMansourAwaed** [comment]
this is how it looks like with new query 

![image](https://github.com/user-attachments/assets/9c81fbd9-bfea-45c3-9dd4-4f61409a6d90)

**Comment by DavidWiseman** [comment]
Those numbers are looking about right now - adding up to 100 or almost 100, but never exceeding it.  Most of the CPU consumption is SQL with a very small amount of CPU consumed by other processes.

**Comment by AhmadMansourAwaed** [comment]
yes the excel file which i uploaded earlier if you it was exceeding the 100 %

with the new script you provided to me it's equal to 100% which is more realistic

**Comment by chadbaldwin** [comment]
@AhmadMansourAwaed I'm just a curious bystander. What do you get for this query:

```sql
SELECT virtual_machine_type_desc FROM sys.dm_os_sys_info
```

??

**Comment by DavidWiseman** [comment]
In addition to the query above, it would also be useful to know what this returns on each of the instances.  

```
/* Get number of numa nodes for the SQL instance */
SELECT COUNT(*) 
FROM sys.dm_os_memory_nodes 
WHERE memory_node_id <> 64
```

**Comment by AhmadMansourAwaed** [comment]
> ```sql
> SELECT virtual_machine_type_desc FROM sys.dm_os_sys_info
> ```

![image](https://github.com/user-attachments/assets/83665c8b-06dd-402d-964e-389d3f739552)

**Comment by AhmadMansourAwaed** [comment]
> /* Get number of numa nodes for the SQL instance */
> SELECT COUNT(*) 
> FROM sys.dm_os_memory_nodes 
> WHERE memory_node_id <> 64

![image](https://github.com/user-attachments/assets/409cfa28-7fa2-41e0-ad6d-f5f726a46c18)

**Comment by DavidWiseman** [comment]
I'll probably have a PR to fix this today or tomorrow and run a custom build for you to try out.  This is useful information you've provided as it confirms the issue can happen on VMs and not just bare metal.

The new collection query will look something like this:

```
DECLARE @TOP INT = 120

DECLARE @ts_now BIGINT 
DECLARE @numa_nodes INT

SELECT	@ts_now= cpu_ticks/(cpu_ticks/ms_ticks)
FROM sys.dm_os_sys_info; 

/* Get number of numa nodes for the SQL instance */
SELECT @numa_nodes  = COUNT(*) 
FROM sys.dm_os_memory_nodes 
WHERE memory_node_id <> 64 /* exclude the internal node for the DAC */

SELECT TOP(@TOP) DATEADD(ms, -1 * (@ts_now - RB.timestamp), GETUTCDATE()) AS EventTime,
				CASE	WHEN RB.SQLProcessUtilization+RB.SystemIdle <= 100 THEN RB.SQLProcessUtilization 
						WHEN (RB.SQLProcessUtilization / @numa_nodes) + RB.SystemIdle <= 100 THEN RB.SQLProcessUtilization / @numa_nodes
						ELSE 100-RB.SystemIdle END AS SQLProcessCPU, 
				RB.SystemIdle AS [SystemIdleProcess] 
FROM (SELECT	record.value('(./Record/@id)[1]', 'int') AS record_id, 
				record.value('(./Record//SystemHealth/SystemIdle)[1]', 'int') AS SystemIdle, 
				record.value('(./Record//SystemHealth/ProcessUtilization)[1]', 'int') AS SQLProcessUtilization, 
				timestamp 
		FROM (SELECT [timestamp], CONVERT(xml, record) AS [record] 
		FROM sys.dm_os_ring_buffers WITH (NOLOCK)
		WHERE ring_buffer_type = N'RING_BUFFER_SCHEDULER_MONITOR' 
		AND record LIKE N'%<SystemHealth>%') AS x) AS RB 
ORDER BY RB.record_id DESC 
OPTION (RECOMPILE);
```

Based off the information @chadbaldwin provided, the calculation needs to be based on physical NUMA nodes rather than processors and we still need to account for the possibility that the SQL process + idle might add up to slightly more than 100%.  I believe the idle is correct so we can do 100-idle to account for this.  The solution isn't perfect with a small risk that the balance of other/SQL process might be incorrect in some situations, but the overall CPU usage reported should be good.

**Comment by DavidWiseman** [comment]
Some additional info on this issue.  The initial fix dividing by physical CPUs doesn't work in all cases.   It fails for this example that has more NUMA nodes than physical CPUs.

![image](https://github.com/user-attachments/assets/c8da1605-9ec9-477a-8cf8-cd6e68c0cf94)

![image](https://github.com/user-attachments/assets/8d94d8e9-0040-4c4e-9151-a9bcdf198366)

Even dividing by numa nodes, we can still add up to slightly more than 100% 

![image](https://github.com/user-attachments/assets/c7616854-2e27-423b-82e0-91f92b26f22e)

Given the idle seems to be correct, we can calculate SQL process as 100-idle if we exceed 100% CPU.  

References:
[Example dividing by NUMA nodes](https://github.com/SparkhoundSQL/sql-server-toolbox/blob/master/cpu%20utilization.sql)
[Comment on Tech Community](https://techcommunity.microsoft.com/t5/sql-server-blog/sql-server-cpu-usage-available-in-sys-dm-os-ring-buffers-dmv/bc-p/4226569)
[DPA fix for this issue](https://support.solarwinds.com/SuccessCenter/s/article/CPU-utilization-is-not-displayed-properly-in-DPA)

Some official guidance/documentation on this issue would be useful.  If anyone has any additional insight or can improve on this solution, let me know.

**Comment by DavidWiseman** [comment]
@AhmadMansourAwaed 
A custom build with the fix is available [here](https://github.com/DavidWiseman/dba-dash/releases/tag/3.15.0)

Instructions for applying the update manually are available [here](https://dbadash.com/docs/setup/upgrades/#manual-upgrade)

**Comment by AhmadMansourAwaed** [comment]
Good morning David , thanks a lot and i wanted to share a good news that i have tried and graph is showing now :)

![image](https://github.com/user-attachments/assets/7b88fdfd-c2dc-42dd-ba85-9178cb6ddc2a)

Thanks again for your support and for the wonderful tool

**Comment by DavidWiseman** [comment]
That's great.  Thanks for letting me know it's working.

**Comment by DavidWiseman** [comment]
Some stats shared from one user on hardware configurations impacted/not impacted by this issue for reference.
![image](https://github.com/user-attachments/assets/a28ea05f-9f62-4c15-bfcb-c40520dcb95f)

**Comment by DavidWiseman** [comment]
Fix included in 3.16 🚀
